README.md
# RStata

[![Linux Build Status](https://travis-ci.org/lbraglia/RStata.svg?branch=master)](https://travis-ci.org/lbraglia/RStata)
[![](http://www.r-pkg.org/badges/version/RStata)](http://www.r-pkg.org/pkg/RStata)
[![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/RStata)](http://www.r-pkg.org/pkg/RStata)

A simple interface for `R -> Stata`, allowing the user to:
- execute Stata commands (both inline and from a `.do` file) from R;
- pass a `data.frame` to Stata;
- return a `data.frame` to R;
- include Stata 'chunks' in knitr documents (however each chunk is
  independent from the others;  feature tested with LaTeX/Linux).

## Install
```
install.packages('RStata')
```

## Setup
In order to use this package you need to setup two `options`:
```
options("RStata.StataPath")
options("RStata.StataVersion")
```
Once done, the main function in the package is `stata` (see Examples below).


### RStata.StataPath
`RStata.StataPath` contains the path to Stata executable (Windows
user have to delete the extension `.exe` from the path ).

You can use

```
chooseStataBin()
```

to set this option the first time:
- in Linux it searches for the 4 (cli) Stata binaries (`stata-mp`,
  `stata-se`, `stata`, `stata-sm`) in the search path and prompt a menu to
  choose one;
- in Windows does a `file.choose` (and a bit of manipulation, 
  in Windows, to obtain a valid path).
- other system are not handled by `chooseStataBin` (mainly because of
  my ignorance regarding both the system itself and Stata installation
  on it, but contributions are welcome).

  **Mac users** are a notable group: at this stage they should set the
  path for the *command line* executable (not the GUI app). Eg for a
  Stata13 SE installation in El Capitain (with default options) the
  path is `'/Applications/Stata/StataSE.app/Contents/MacOS/stata-se'`.  In
  other cases (Stata licenses, installation path and so on) the path
  could be slightly different (BTW, i've found the path 'exploring'
  the filesystem with a terminal, not the graphics explorer since it
  'stops' at 'StataSE.app').
  

`chooseStataBin` set up `RStata.StataPath` accordingly to user advice and
return the path; but you may consider to put the path returned in
`.Rprofile` options for the next time you start R.

### RStata.StataVersion
`RStata.StataVersion` is needed for right management of data export/import
to/from Stata.  Eg, for Stata 13:
```
options("RStata.StataVersion" = 13)
```

As above, you may consider to put this statement in your `.Rprofile`.

### Other options
See `?stata`.

## Examples

### One inline command
```
> stata("help regress")  #<- this won't work in Windows dued to needed
                         #   batch mode

. help regress

Title
-----

    [R] regress -- Linear regression


Syntax
------

        regress depvar [indepvars] [if] [in] [weight] [, options]

[...]
```

### Many inline commands
From version 0.4.0 you can do this (eg.):
```
> stata_src <- "

version 10
sysuse auto
reg mpg weight

"
> stata(stata_src)

[...]
```

### External .do file sourceing
```
> stata("foo.do")
```

### Data input to Stata
```
> x <- data.frame(a = rnorm(3), b = letters[1:3])
> stata("sum a", data.in = x)

. sum a

   Variable |       Obs        Mean    Std. Dev.       Min        Max
------------+--------------------------------------------------------
          a |         3   -.5603985    .3000552  -.7720861  -.2170166
```

### Data output from Stata
Eg in order to obtain `auto` dataset
```
> auto <- stata("sysuse auto", data.out = TRUE)

. sysuse auto
(1978 Automobile Data)
 
> head(auto)
           make price mpg rep78 headroom trunk weight length turn displacement
1   AMC Concord  4099  22     3      2.5    11   2930    186   40          121
2     AMC Pacer  4749  17     3      3.0    11   3350    173   40          258
3    AMC Spirit  3799  22    NA      3.0    12   2640    168   35          121
4 Buick Century  4816  20     3      4.5    16   3250    196   40          196
5 Buick Electra  7827  15     4      4.0    20   4080    222   43          350
6 Buick LeSabre  5788  18     3      4.0    21   3670    218   43          231
  gear_ratio  foreign
1       3.58 Domestic
2       2.53 Domestic
3       3.08 Domestic
4       2.93 Domestic
5       2.41 Domestic
6       2.73 Domestic
```

### Data input/output
```
> x <- data.frame(a = rnorm(3), b = letters[1:3])
> (y <- stata("replace a = 2", data.in = x, data.out = TRUE))

. replace a = 2
(3 real changes made)

  a b
1 2 a
2 2 b
3 2 c
```


## News
[Here](http://raw.githubusercontent.com/lbraglia/RStata/master/NEWS).

## Contribute
You are welcome to:
* submit suggestions and bug-reports at: <http://github.com/lbraglia/RStata/issues>
* fork and send a pull request on: <http://github.com/lbraglia/RStata/>
* send an e-mail to: <lbraglia@gmail.com>


NEWS
RStata 1.1.0
----------------------------------------------------------------

BUG FIXES
* chooseStataBin: fixed quoting in Stata path on Windows
* stata: removed unused parameter stata.quiet
* misc code cleaning

RStata 1.0.0
----------------------------------------------------------------

NEW FEATURES
* Available on CRAN


RStata 0.4.1
----------------------------------------------------------------

BUG FIXES
* cleaner output and removed unprintable chars (at the beginning of the
  output). Now it's possibile to have knitr document including Stata
  'chunk'(tested on Linux).


RStata 0.4.0
----------------------------------------------------------------

NEW FEATURES
* Now it is possibile to code multiple statements in a more natural way
  (due to a bit of strsplit), eg:

  ```
  stata_src <- '

  version 10
  set more off
  sysuse auto
  reg mpg weight

  '
  stata(stata_src)
  ```    


RStata 0.3.0
----------------------------------------------------------------

BUG FIXES
* capture noisily by default in order to avoid unneeded freezings (eg when
  making mistakes with stata syntax).
  

RStata 0.2.2
----------------------------------------------------------------

NEW FEATURES
* Added read/write compatibility for Stata 14 (#3). Thanks to tystan.

BUG FIXES
* set more off by default at the beginning of the script, in order to avoid
  unneeded freezings (eg when looking documentation).


RStata 0.2.1
----------------------------------------------------------------

BUG FIXES
* better README.md.


RStata 0.2.0
----------------------------------------------------------------

NEW FEATURES

* NEWS introduced;
* x.x.x numbering introduced.


R/chooseStataBin.R
#' Set Stata binary path
#'
#' Set Stata binary (among found alternatives) path. These settings are
#' lost when R is closed, therefore you should consider adding a
#' \code{options("RStata.StataPath")} line in your \code{.Rprofile}.
#' 
#' @export
chooseStataBin <- function()
{
    OS <- Sys.info()["sysname"]
    OS.type <- .Platform$OS.type
  
    ## ------------------------------
    if (OS %in% "Linux") {
        m <- c(`Stata MP` = "stata-mp",
               `Stata SE` = "stata-se",
               `Stata IC` = "stata",
               `Small Stata` = "stata-sm" )
        
        bin <- Sys.which(m)
        names(bin) <- names(m)
        nApps <- length(availProg <- bin[ "" != bin])
    
        if (0 == nApps) {
            stop("No application (detected) availables.\n",
                 "Set options('RStata.StataPath'), instead." )

        } else if (1 == nApps) {
            
            cat("Only", names(availProg), "found; I'll use it.\n")
            unnprog <- unname(availProg)
            options(RStata.StataPath = unnprog)
            return(unnprog)

        } else if (nApps > 1) {

            if (!interactive())
                stop("Cannot choose a Stata bin path non-interactively.\n",
                     "Set options('RStata.StataPath'), instead.")
            res <- utils::menu(names(availProg), title = "Stata availables")
            if (res > 0L) {
                unnprog <- unname(availProg[res])
                options(RStata.StataPath = unnprog)
                return(unnprog)
            }

        } else {
            stop("Unexpected error")
        }
        ## ------------------------------
    } else if (OS %in% "Windows"){
        prog <- file.choose()
        prog <- shQuote(tools::file_path_sans_ext(prog))
        options(RStata.StataPath = prog)
        return(prog)
    } else {
        ""
    }
    
}


R/stata.R
#' Send commands to a Stata process
#'
#' Function that sends commands to a Stata process.
#' @param src character vector of length 1 (path to \code{.do} file) or more
#' (a set of stata commands). See examples.
#' @param data.in \code{\link{data.frame}} to be passed to Stata
#' @param data.out logical value. If \code{TRUE}, the data at the end of
#' the Stata command are returned to R.
#' @param stata.path Stata command to be used
#' @param stata.version Version of Stata used
#' @param stata.echo logical value. If \code{TRUE} stata text output will be printed
#' @param ... parameter passed to \code{\link{write.dta}}
#' @examples
#' \dontrun{
#' ## Single command
#' stata("help regress") #<- this won't work in Windows dued to needed
#'                       #   batch mode
#'
#' ## Many commands
#' stata_src <- '
#'
#' version 10
#' set more off
#' sysuse auto
#' reg mpg weight
#'
#' '
#' stata(stata_src)
#'
#' ## External .do file
#' stata("foo.do")
#'
#' ## Data input to Stata
#' x <- data.frame(a = rnorm(3), b = letters[1:3])
#' stata( "sum a", data.in = x)
#'
#' ## Data output from Stata (eg obtain 'auto' dataset)
#' auto <- stata("sysuse auto", data.out = TRUE)
#' head(auto)
#'
#' ## Data input/output
#' (y <- stata("replace a = 2", data.in = x, data.out = TRUE))
#' }
#' @export
stata <- function(src = stop("At least 'src' must be specified"),
                  data.in = NULL,
                  data.out = FALSE,
                  stata.path = getOption("RStata.StataPath", stop("You need to set up a Stata path; ?chooseStataBin")),
                  stata.version = getOption("RStata.StataVersion", stop("You need to specify your Stata version")),
                  stata.echo = getOption("RStata.StataEcho", TRUE),
                  ...
                  )
{
    ## -------------------------
    ## Data validation and setup
    ## -------------------------
    if (!is.character(src))
        stop("src must be a character")
  
    if (!(is.null(data.in) | is.data.frame(data.in)))
        stop("data.in must be NULL or a data.frame")
  
    if (!is.logical(data.out))
        stop("data.out must be logical")

    if (!is.numeric(stata.version))
        stop("stata.version must be logical")

    if (!is.logical(stata.echo))
        stop("stata.echo must be logical")

    OS <- Sys.info()["sysname"]
    OS.type <- .Platform$OS.type
    SRC <- unlist(lapply(src, strsplit, '\n'))
    dataIn <- is.data.frame(data.in)
    dataOut <- data.out[1L]
    stataVersion <- stata.version[1L]
    stataEcho <- stata.echo[1L]

    ## -----------------
    ## OS related config
    ## -----------------
    ## in Windows and batch mode a RStata.log (naming after RStata.do
    ## below) is generated in the current directory

    if (OS %in% "Windows") {
        winRStataLog <- "RStata.log"
        on.exit(unlink(winRStataLog))
    }
  
    ## -----
    ## Files
    ## -----

    ## tempfile could be misleading if the do source other dos 
    ## with relative paths
    doFile <- "RStata.do"
    on.exit(unlink(doFile), add = TRUE)

    if (dataIn){
        ## dtaInFile <- tempfile("RStataDataIn", fileext = ".dta")
        ## Windows/Stata8 unhappy?
        dtaInFile <- "RStataDataIn.dta"
        on.exit(unlink(dtaInFile), add = TRUE)
        foreign::write.dta(data.in, file = dtaInFile,
                           version = ifelse(stataVersion >= 7, 7L, 6L), ...)
    }

    if (dataOut) {
        ## dtaOutFile <- tempfile("RStataDataOut", fileext = ".dta")
        ## Windows/Stata8 unhappy?
        dtaOutFile <- "RStataDataOut.dta"
        on.exit(unlink(dtaOutFile), add = TRUE)
    }

    ## -------------------------
    ## Creating the .do file ...
    ## -------------------------
    
    ## External .do script 'support': KIS
    if (file.exists(SRC[1L]))
        SRC <- readLines(SRC[1L])

    ## put a placeholder around the part of interest, in order to find
    ## it easily (when removing overhead/setup code for each run)
    cut_me_here <- 'RSTATA: cut me here'
    cut_me_comment <- paste0('/*', cut_me_here, '*/')

    ## capture noisily and set cut points
    SRC <- c(
    {if (dataIn) sprintf("use %s", tools::file_path_sans_ext(dtaInFile))
     else ''},
    'capture noisily {',
    cut_me_comment,
    SRC,
    cut_me_comment,
    '} /* end capture noisily */')

    ## set more off just to be sure nothing will freeze (hopefully :) )
    SRC <- c('set more off', SRC)
    
    ## put a save or saveold at the end of .do if data.out == TRUE
    ## for Stata 14, saveold defaults to a Stata 13 dta file
    ## -> use the (Stata 14 only) saveold option: "version(12)" to allow
    ## foreign::read.dta() read compatibility
    if (dataOut)
        SRC <- c(SRC, sprintf("%s %s%s",
                              ifelse(stataVersion >= 13, "saveold", "save"),
                              tools::file_path_sans_ext(dtaOutFile),
                              ifelse(stataVersion >= 14, ", version(12)", "")))
    
    ## adding this command to the end simplify life if user make changes but
    ## doesn't want a data.frame back
    SRC <- c(SRC, "exit, clear STATA")

    ## -------------
    ## Stata command
    ## -------------

    ## With Windows version, /e is almost always needed (if Stata is
    ## installed with GUI)
    stataCmd <- paste(stata.path,
                      ifelse(OS %in% "Windows", "/e", ""),
                      "do",
                      doFile)
  
    ## ---
    ## IPC
    ## ---
    ## setup the .do file
    ## con <- fifo(doFile, "w+") # <- freeze with fifo in Window
    con <- file(doFile, "w")
    writeLines(SRC, con)
    close(con)
    
    ## execute Stata
    rdl <- pipe(stataCmd, "r")
    stataLog <- readLines(rdl)
    close(rdl)
    
    if (stataEcho) {
        if (OS %in% "Windows") stataLog <- readLines(winRStataLog)
        ## postprocess log, keeping only the output of interest (between rows
        ## having /* RSTATA: cut me here */
        cutpoints <- grep(cut_me_here, stataLog)
        stataLog <- stataLog[seq.int(cutpoints[1] + 1, cutpoints[2] - 1)]
        cat(stataLog, sep = "\n")
    }

    ## ------------------
    ## Get data outputted
    ## ------------------
    if (dataOut){
        res <- foreign::read.dta(dtaOutFile, ...)
        invisible(res)
    }
  
}
